src/interactions/brush.js

```diff
@@ -0,0 +1,138 @@
+              ? []
+              : mode === "xy"
+              ? transpose(selection)
+              : mode === "x"
+              ? [selection]
+              : [, selection];
+            if (X || Y) S = d.index;
+            if (X) {
+              const [x0, x1] = X;
+              S = S.filter((i) => x0 <= Xm[i] && Xl[i] <= x1);
+            }
+            if (Y) {
+              const [y0, y1] = Y;
+              S = S.filter((i) => y0 <= Ym[i] && Yl[i] <= y1);
+            }
+            // Only one facet can be active at a time; clear the others.
+            if (type === "start") for (const target of targets) if (target !== this) target._cancelBrush();
+
+            d.display?.replaceWith((d.display = next.call(this, S ?? [], scales, values, dimensions, context)));
+
+            // Update the plot’s value if the selection has changed.
+            if (selectionMode === "data") {
+              if (!selectionEquals(S, state.selection)) {
+                state.selection = S;
+                context.dispatchValue(S === null ? data : take(data, S));
+              }
+            }
+            // "extent"
+            else {
+              if (selection === null) {
+                context.dispatchValue(null);
+              } else {
+                const value = {};
+                if (X) value.x = x.invert ? X.map(x.invert).sort(ascending) : X;
+                if (Y) value.y = y.invert ? Y.map(y.invert).sort(ascending) : Y;
+                if ("fx" in scales) value.fx = index.fx;
+                if ("fy" in scales) value.fy = index.fy;
+                context.dispatchValue(value);
+              }
+            }
+          });
+
+        states.set(svg, (state = {brush, targets: [], selection: null}));
+      }
+      const {brush, targets} = state;
+      const display = next.call(this, [], scales, values, dimensions, context);
+
+      // Create a wrapper for the elements to display, and a target that will
+      // carry the brush. Save references to the display and index of the current
+      // facet.
+      const wrapper = create("svg:g", context);
+      const target = create("svg:g", context).attr("aria-label", "brush").datum({display, index}).call(brush);
+      const node = target.node();
+      node._cancelBrush = () => target.call(brush.move, null);
+      targets.push(node);
+
+      if (display) {
+        const al = display.getAttribute("aria-label");
+        al && (wrapper.attr("aria-label", al), display.removeAttribute("aria-label"));
+        wrapper.append(() => display);
+      }
+
+      // Translate the brush target to match the facet’s frame position, and
+      // initialize the plot’s value. We skip a beat so all the marks have been
+      // added before the brush (to top z-index).
+      Promise.resolve().then(() => {
+        node.setAttribute("transform", wrapper.attr("transform"));
+        svg.appendChild(node);
+        context.dispatchValue(selectionMode === "data" ? data : null);
```

Comment:
It’s not ideal to dispatch an _input_ event on mount since the user didn’t really interact with the plot. If it were synchronous, it’d be okay since the element by definition wouldn’t be attached to the DOM and thus the event wouldn’t go anywhere. But here it’s asynchronous, so it could trigger redundant computation. The chart’s value should ideally be set synchronously during render so that it’s resolved when render returns. (I think the pointer transform has this problem too, but only when the chart is faceted… there’s probably a way we could fix that.)

I also don’t like making this asynchronous as a way of drawing the brush on top. That feels like it violates our normal way of controlling z-order through the order of the **marks** option; what if there are other things that want to be “on top” in the future, too. Like the tip mark? Or some other interaction we haven’t designed yet? Perhaps we can avoid this for now by only applying the brush to the last mark, or perhaps we need to provide a separate explicit way of declaring marks that go on top.